File: Codeforces/E_XOR_and_Favorite_Number.py
import sys
input = sys.stdin.readline
from collections import defaultdict

n, m, k = map(int, input().split())
arr = list(map(int, input().split()))

pxor = [0]
x = 0
for num in arr:
    x ^= num
    pxor.append(x)

all_vals = set(pxor)
for val in pxor:
    all_vals.add(val ^ k)

compressed = {val: idx for idx, val in enumerate(sorted(all_vals))}
pxor = [compressed[val] for val in pxor]

BLOCK = int(n ** 0.5)
queries = []
for i in range(m):
    l, r = map(int, input().split())
    queries.append((l, r, i))

queries.sort(key=lambda x: ((x[0] // BLOCK, x[1]) if (x[0] // BLOCK) % 2 == 0 else (x[0] // BLOCK, -x[1])))

cnt = 0
answers = [0] * m
freq = [0] * (2 * n + 10)
curr_l = 1
curr_r = 0
freq[pxor[0]] = 1

for l, r, idx in queries:
    while curr_r < r:
        curr_r += 1
        val = pxor[curr_r]
        cnt += freq[val ^ compressed[k]]
        freq[val] += 1

    while curr_r > r:
        val = pxor[curr_r]
        freq[val] -= 1
        cnt -= freq[val ^ compressed[k]]
        curr_r -= 1

    while curr_l < l:
        val = pxor[curr_l - 1]
        freq[val] -= 1
        cnt -= freq[val ^ compressed[k]]
        curr_l += 1

    while curr_l > l:
        curr_l -= 1
        val = pxor[curr_l - 1]
        cnt += freq[val ^ compressed[k]]
        freq[val] += 1

    answers[idx] = cnt

print('\n'.join(map(str, answers)))

File: Codeforces/E_XOR_and_Favorite_Number.cpp
#include <bits/stdc++.h>
using namespace std;

const int MAXN = 1e6 + 5;
int n, m, k;
vector<int> arr;
vector<int> prefix;
vector<tuple<int, int, int>> queries;
vector<long long> answers;
int BLOCK;

bool mo_cmp(const tuple<int, int, int> &a, const tuple<int, int, int> &b) {
    int block_a = get<0>(a) / BLOCK;
    int block_b = get<0>(b) / BLOCK;
    if (block_a != block_b) return block_a < block_b;
    return get<1>(a) < get<1>(b);
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);

    cin >> n >> m >> k;
    arr.resize(n);
    prefix.resize(n + 1);
    answers.resize(m);

    for (int i = 0; i < n; ++i) cin >> arr[i];

    prefix[0] = 0;
    for (int i = 0; i < n; ++i) prefix[i + 1] = prefix[i] ^ arr[i];

    BLOCK = sqrt(n) + 1;

    for (int i = 0; i < m; ++i) {
        int l, r;
        cin >> l >> r;
        queries.emplace_back(l, r, i);
    }

    sort(queries.begin(), queries.end(), mo_cmp);

    unordered_map<int, int> freq;
    freq[0] = 1;

    int curr_l = 1, curr_r = 0;
    long long cnt = 0;

    for (auto &[l, r, idx] : queries) {
        while (curr_r < r) {
            ++curr_r;
            int val = prefix[curr_r];
            cnt += freq[val ^ k];
            ++freq[val];
        }
        while (curr_r > r) {
            int val = prefix[curr_r];
            --freq[val];
            cnt -= freq[val ^ k];
            --curr_r;
        }
        while (curr_l < l) {
            int val = prefix[curr_l - 1];
            --freq[val];
            cnt -= freq[val ^ k];
            ++curr_l;
        }
        while (curr_l > l) {
            --curr_l;
            int val = prefix[curr_l - 1];
            cnt += freq[val ^ k];
            ++freq[val];
        }

        answers[idx] = cnt;
    }

    for (auto ans : answers) cout << ans << '\n';

    return 0;
}
